Upstream: http://lists.gnu.org/archive/html/grub-devel/2009-09/msg00056.html
Description: Don't display introductory message about line editing unless
 we're actually offering a shell prompt. Also, don't clear the screen just
 before booting if we never drew the menu in the first place.
UbuntuSpecific: The line editing change is believed to be a workaround for a
 different bug. We'll go with this for now, but will drop this in favour of
 a better fix upstream if somebody figures out what that is.


diff -ur grub2-1.98-ylmf2/normal/main.c grub2-1.98-ylmf2.new/normal/main.c
--- grub2-1.98-ylmf2/normal/main.c	2010-03-07 04:51:37.000000000 +0800
+++ grub2-1.98-ylmf2.new/normal/main.c	2011-05-25 16:31:34.529540944 +0800
@@ -566,6 +566,14 @@
   grub_parser_t parser = grub_parser_get_current ();
   char *prompt;
 
+  static int displayed_intro;
+
+  if (! displayed_intro)
+    {
+      grub_normal_reader_init (nested);
+      displayed_intro = 1;
+    }
+
   if (cont)
     prompt = grub_xasprintf (">");
   else
@@ -614,8 +622,6 @@
       return;
     }
 
-  grub_normal_reader_init (nested);
-
   while (1)
     {
       char *line;
diff -ur grub2-1.98-ylmf2/normal/menu.c grub2-1.98-ylmf2.new/normal/menu.c
--- grub2-1.98-ylmf2/normal/menu.c	2010-03-07 04:51:37.000000000 +0800
+++ grub2-1.98-ylmf2.new/normal/menu.c	2011-05-25 16:32:32.862873505 +0800
@@ -542,6 +542,7 @@
       int boot_entry;
       grub_menu_entry_t e;
       int auto_boot;
+      int initial_timeout = grub_menu_get_timeout ();
 
       boot_entry = run_menu (menu, nested, &auto_boot);
       if (boot_entry < 0)
@@ -551,7 +552,9 @@
       if (! e)
 	continue; /* Menu is empty.  */
 
-      grub_cls ();
+       /* Only clear the screen if we drew the menu in the first place.  */
+       if (initial_timeout != 0)
+        grub_cls ();
 
       if (auto_boot)
         {
